### PR TITLE
Add support for xor_iv

### PR DESCRIPTION
**Pull request by huitema**
Generalize the proof of concept to openssl, minicrypto, fusion and bcrypt.
Add unit tests.

### CODE REVIEW

File: include/picotls.h [hunk 1]
```diff
@@ -1228,6 +1229,11 @@ ptls_aead_context_t *ptls_aead_new_direct(ptls_aead_algorithm_t *aead, int is_en
  * destroys an AEAD cipher context
  */
 void ptls_aead_free(ptls_aead_context_t *ctx);
+/**
+ * Permutates the static IV by applying given bytes using bit-wise XOR. This API can be used for supplying nonces longer than 64-
```
Comment (kazuho) [inline include/picotls.h:1233]:
```suggestion
 * Permutes the static IV by applying given bytes using bit-wise XOR. This API can be used for supplying nonces longer than 64-
```

I know that this is my typo but appreciate it being fixed.

File: lib/cifra/aes-common.h [hunk 2]
```diff
@@ -154,11 +154,21 @@ static inline size_t aesgcm_decrypt(ptls_aead_context_t *_ctx, void *output, con
...
+    for (size_t i = 0; i < len; ++i)
+        ctx->static_iv[i] ^= bytes[i];
+}
+
 static inline int aead_aesgcm_setup_crypto(ptls_aead_context_t *_ctx, int is_enc, const void *key, const void *iv)
 {
     struct aesgcm_context_t *ctx = (struct aesgcm_context_t *)_ctx;
 
     ctx->super.dispose_crypto = aesgcm_dispose_crypto;
+    ctx->super.do_xor_iv = minicrypto_aesgcm_aead_xor_iv;
```
Comment (kazuho) [inline lib/cifra/aes-common.h:171]:
```suggestion
    ctx->super.do_xor_iv = aesgcm_xor_iv;
```

File: lib/cifra/aes-common.h [hunk 3]
```diff
@@ -154,11 +154,21 @@ static inline size_t aesgcm_decrypt(ptls_aead_context_t *_ctx, void *output, con
     return tag_offset;
 }
 
+static inline void minicrypto_aesgcm_aead_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```
Comment (kazuho) [inline lib/cifra/aes-common.h:157]:
```suggestion
static inline void aesgcm_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```

Let's follow the current naming convention.

File: lib/cifra/chacha20.c [hunk 4]
```diff
@@ -179,11 +179,21 @@ static size_t chacha20poly1305_decrypt(ptls_aead_context_t *_ctx, void *output,
     return ret;
 }
 
+static void minicrypto_chacha20_aead_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```
Comment (kazuho) [inline lib/cifra/chacha20.c:182]:
```suggestion
static void chacha20poly1305_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```

File: lib/cifra/chacha20.c [hunk 5]
```diff
@@ -179,11 +179,21 @@ static size_t chacha20poly1305_decrypt(ptls_aead_context_t *_ctx, void *output,
...
+    for (size_t i = 0; i < len; ++i)
+        ctx->static_iv[i] ^= bytes[i];
+}
+
 static int aead_chacha20poly1305_setup_crypto(ptls_aead_context_t *_ctx, int is_enc, const void *key, const void *iv)
 {
     struct chacha20poly1305_context_t *ctx = (struct chacha20poly1305_context_t *)_ctx;
 
     ctx->super.dispose_crypto = chacha20poly1305_dispose_crypto;
+    ctx->super.do_xor_iv = minicrypto_chacha20_aead_xor_iv;
```
Comment (kazuho) [inline lib/cifra/chacha20.c:196]:
```suggestion
    ctx->super.do_xor_iv = chacha20poly1305_xor_iv;
```

File: lib/fusion.c [hunk 6]
```diff
@@ -934,6 +946,7 @@ static int aesgcm_setup(ptls_aead_context_t *_ctx, int is_enc, const void *key,
         return 0;
 
     ctx->super.dispose_crypto = aesgcm_dispose_crypto;
+    ctx->super.do_xor_iv = aesgcm_fusion_aead_xor_iv;
```
Comment (kazuho) [inline lib/fusion.c:949]:
```suggestion
    ctx->super.do_xor_iv = aesgcm_xor_iv;
```

File: lib/fusion.c [hunk 7]
```diff
@@ -924,6 +928,14 @@ static size_t aead_do_decrypt(ptls_aead_context_t *_ctx, void *output, const voi
     return enclen;
 }
 
+static inline void aesgcm_fusion_aead_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```
Comment (kazuho) [inline lib/fusion.c:931]:
```suggestion
static inline void aesgcm_xor_iv(ptls_aead_context_t *_ctx, const void *_bytes, size_t len)
```

File: t/fusion.c [hunk 8]
```diff
@@ -19,6 +19,14 @@
...
  * IN THE SOFTWARE.
  */
+
+#if defined(_WINDOWS) && !defined(_WINDOWS64)
+#include <stdio.h>
+int main(int argc, char **argv)
+{
+    printf("Fusion is disabled on x86 32 bits builds.\n");
+}
+#else
```
Comment (kazuho) [inline t/fusion.c:29]: I think we should just skip building (or testing) fusion on platforms that do not support it. FWIW, we do that for t/openssl.c too.

**Review by kazuho — COMMENTED** [review]
Thank you for the PR. Looking good, mostly editorial comments.

File: t/fusion.c (hunk 8 above)
Comment (huitema) [reply t/fusion.c:29]: I have to find the right way to do that with VS...

Comment (huitema) [reply t/fusion.c]: Found the issue. I had duplicated the fusion test in VS...

Comment (huitema) [reply t/fusion.c]: Duplicated twice, got confused in Visual Studio. All cleaned up now.

**Pull request merged**